WooCommerce/Classes/ViewRelated/Feature Highlight/Tooltip.swift

```diff
@@ -0,0 +1,420 @@
+            ),
+            containerView.widthAnchor.constraint(
+                lessThanOrEqualToConstant: maxWidth
+            ),
+            buttonsStackView.heightAnchor.constraint(equalToConstant: Constants.Spacing.buttonStackViewHeight)
+        ])
+    }
+
+    private func addShadow() {
+        containerView.layer.shadowColor = UIColor.black.cgColor
+        containerView.layer.shadowOffset = CGSize(width: 0, height: 2)
+        containerView.layer.shadowOpacity = traitCollection.userInterfaceStyle == .light ? 0.5 : 0
+    }
+
+    @objc private func didTapPrimaryButton() {
+        dismissalAction?()
+    }
+
+    @objc private func didTapSecondaryButton() {
+        secondaryButtonAction?()
+    }
+
+    private static func height(
+        withTitle title: String?,
+        message: String?
+    ) -> CGFloat {
+        var totalHeight: CGFloat = 0
+
+        totalHeight += Constants.Spacing.contentStackViewTop
+
+        if let title = title {
+            totalHeight += title.height(withMaxWidth: maxContentWidth(), font: UIFont.body)
+        }
+
+        totalHeight += Constants.Spacing.contentStackViewInterItemSpacing * 2
+
+        if let message = message {
+            totalHeight += message.height(withMaxWidth: maxContentWidth(), font: UIFont.body)
+        }
+
+        totalHeight += Constants.Spacing.buttonStackViewHeight
+        totalHeight += Constants.Spacing.contentStackViewBottom
+
+        return totalHeight
+    }
+
+    private static func width(
+        title: String?,
+        message: String?,
+        primaryButtonTitle: String?,
+        secondaryButtonTitle: String?
+    ) -> CGFloat {
+        let titleWidth = title?.width(withMaxWidth: maxContentWidth(), font: UIFont.body) ?? 0
+        let messageWidth = message?.width(withMaxWidth: maxContentWidth(), font: UIFont.body) ?? 0
+
+        var buttonsWidth: CGFloat = 0
+        if let primaryButtonTitle = primaryButtonTitle {
+            buttonsWidth += primaryButtonTitle.width(withMaxWidth: maxContentWidth(), font: UIFont.subheadline)
+        }
+
+        if let secondaryButtonTitle = secondaryButtonTitle {
+            buttonsWidth += secondaryButtonTitle.width(
+                withMaxWidth: maxContentWidth(),
+                font: UIFont.subheadline
+            ) + Constants.Spacing.buttonsStackViewInterItemSpacing
+        }
+
+        return max(max(titleWidth, messageWidth), buttonsWidth) + Constants.Spacing.contentStackViewHorizontal * 2
+    }
+
+    private static func maxContentWidth() -> CGFloat {
+        UIScreen.main.bounds.width
+        - Constants.Spacing.superHorizontalMargin
+        - (Constants.Spacing.contentStackViewHorizontal * 2)
+    }
+}
+
+extension String {
+    private func size(withMaxWidth maxWidth: CGFloat, font: UIFont) -> CGRect {
+        let constraintRect = CGSize(width: maxWidth, height: .greatestFiniteMagnitude)
+        let boundingBox = self.boundingRect(
```

Comment:
nit: is `self` needed?
```suggestion
        let boundingBox = boundingRect(
```

Reply:
It looks like self is needed as we are accessing an NSString extension's method.